src/plugins/mer/mervirtualmachinesettingswidget.cpp

```diff
@@ -0,0 +1,107 @@
+#include "mervirtualmachinesettingswidget.h"
+#include "mervirtualboxmanager.h"
+#include <utils/utilsicons.h>
+#include "ui_mervirtualmachinesettingswidget.h"
+
+namespace Mer {
+namespace Internal {
+
+namespace {
+const int DEFAULT_MAX_MEMORY_SIZE_MB = 3584;
+const int MAX_VDI_SIZE_INCREMENT_GB = 10;
+
+} // namespace anonymous
+
+MerVirtualMachineSettingsWidget::MerVirtualMachineSettingsWidget(QWidget *parent)
+    : QWidget(parent)
+    , ui(new Ui::MerVirtualMachineSettingsWidget)
+{
+    ui->setupUi(this);
+    initGui();
+}
+
+MerVirtualMachineSettingsWidget::~MerVirtualMachineSettingsWidget()
+{
+    delete ui;
+}
+
+void MerVirtualMachineSettingsWidget::setMemorySizeMb(int sizeMb)
+{
+    if (ui->memorySpinBox->maximum() < sizeMb)
+        ui->memorySpinBox->setRange(4, sizeMb);
+
+    ui->memorySpinBox->setValue(sizeMb);
+}
+
+void MerVirtualMachineSettingsWidget::setCpuCount(int count)
+{
+    ui->cpuCountSpinBox->setValue(count);
+}
+
+void MerVirtualMachineSettingsWidget::setVdiCapacityMb(int vdiCapacityMb)
+{
+    const double vdiCapacityGb = vdiCapacityMb / 1024.0;
+    // Prohibit adding more than 10GB in one step, because there is no way to go back
+    // the minimum size is the current size because VBoxManager can't reduce storage size
+    ui->vdiCapacityGbSpinBox->setRange(vdiCapacityGb, vdiCapacityGb + MAX_VDI_SIZE_INCREMENT_GB);
+    ui->vdiCapacityGbSpinBox->setValue(vdiCapacityGb);
+}
+
+void MerVirtualMachineSettingsWidget::setVmOff(bool vmOff)
+{
+    ui->memorySpinBox->setEnabled(vmOff);
+    ui->cpuCountSpinBox->setEnabled(vmOff);
+    ui->vdiCapacityGbSpinBox->setEnabled(vmOff);
+
+    ui->memoryInfoLabel->setVisible(!vmOff);
+    ui->cpuInfoLabel->setVisible(!vmOff);
+    ui->vdiCapacityInfoLabel->setVisible(!vmOff);
+}
+
+QFormLayout *MerVirtualMachineSettingsWidget::formLayout() const
+{
+    return ui->formLayout;
+}
+
+void MerVirtualMachineSettingsWidget::initGui()
+{
+    auto handler = [this](int sizeMb) {
+        int maxMemorySizeMb = ui->memorySpinBox->value() > sizeMb ? ui->memorySpinBox->value() : sizeMb;
+        ui->memorySpinBox->setRange(4, maxMemorySizeMb);
+    };
+
+    ui->memorySpinBox->setRange(4, DEFAULT_MAX_MEMORY_SIZE_MB);
+    MerVirtualBoxManager::getHostTotalMemorySizeMb(this, handler);
+
+    const int maxCpuCount = MerVirtualBoxManager::getHostTotalCpuCount();
+    ui->cpuCountSpinBox->setRange(1, maxCpuCount);
+
+    connect(ui->memorySpinBox, SIGNAL(valueChanged(int)),
+            this, SIGNAL(memorySizeChanged(int)));
+    connect(ui->cpuCountSpinBox, SIGNAL(valueChanged(int)),
+            this, SIGNAL(cpuCountChanged(int)));
+    connect(ui->vdiCapacityGbSpinBox, QOverload<double>::of(&QDoubleSpinBox::valueChanged), [this](double value) {
+        emit vdiCapacityChnaged(static_cast<int>(value * 1024));
```

Comment:
This seems error prone. Imagine when `setVdiCapacityMb(1024+512)` is first called, then the Gb spin box is incremented to 2 and returned back to 1 - nothing changed from user point of view, but this will emit `vdiCapacityChanged(1024)`.

Reply:
Hm, i can't repeat error. 
In first call of `setVdiCapacityMb(1024+512)`  will be set range `ui->vdiCapacityGbSpinBox->setRange(1.5, 11.5)`. With this range signal `QDoubleSpinBox::valueChanged` doesn't call  with 1

Reply:
ah, I missed it's a QDoubleSpinBox